core/libp2p/libp2p.hpp

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_LIBP2P_HPP
+#define KAGOME_LIBP2P_HPP
+
+#include <memory>
+
+#include "common/result.hpp"
+#include "libp2p/common/peer_info.hpp"
+#include "libp2p/connection/connection.hpp"
+#include "libp2p/error/error.hpp"
+#include "libp2p/multi/multiaddress.hpp"
+#include "libp2p/multi/multihash.hpp"
+#include "libp2p/routing/router.hpp"
+#include "libp2p/store/record_store.hpp"
+#include "libp2p/swarm/swarm.hpp"
+
+namespace libp2p {
+  /**
+   * Provides a way to instantiate a libp2p instance via joining different
+   * components to it and dial to the peer
+   */
+  class Libp2p {
+    /**
+     * Add a swarm component
+     * @param swarm to be added
+     */
+    virtual void addSwarm(std::unique_ptr<swarm::Swarm> swarm) = 0;
```

Comment:
Can't it be passed again via ctor? There's no sense to call dial without initialization

Reply:
before comments like this you should really go and read https://github.com/libp2p/specs/blob/master/6-interfaces.md :)

Reply:
We need no generic library, but specific to the task. So there's no need in providing complete interface, valuable for everyone

Reply:
We decided to use a spec interface, as it is implemented by two main implementations of Libp2p, thus it should really be reliable. I think, it's better to stick to it at least now, and if some problems appear during the implementation or usage, it is not very hard to change it a little